specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json

```diff
@@ -0,0 +1,1074 @@
+        }
+      ],
+      "properties": {
+        "properties": {
+          "$ref": "#/definitions/AzureFunctionOutputDataSourceProperties",
+          "description": "The properties that are associated with a Azure Function output. Required on PUT (CreateOrReplace) requests.",
+          "x-ms-client-flatten": true
+        }
+      }
+    },
+    "AzureFunctionOutputDataSourceProperties": {
+      "description": "The properties that are associated with a DocumentDB output.",
+      "properties": {
+        "functionAppName": {
+          "description": "The name of your Azure Functions app.",
+          "type": "string"
+        },
+        "functionName": {
+          "description": "The name of the function in your Azure Functions app.",
+          "type": "string"
+        },
+        "apiKey": {
+          "description": "If you want to use an Azure Function from another subscription, you can do so by providing the key to access your function.",
+          "type": "string"
+        },
+        "maxBatchSize": {
+          "description": "A property that lets you set the maximum size for each output batch that's sent to your Azure function. The input unit is in bytes. By default, this value is 262,144 bytes (256 KB).",
+          "type": "number"
+        },
+        "maxBatchCount": {
+          "description": "A property that lets you specify the maximum number of events in each batch that's sent to Azure Functions. The default value is 100.",
+          "type": "number"
+        }
+      }
+    },
+    "ServiceBusQueueOutputDataSource": {
+      "description": "Describes a Service Bus Queue output data source.",
+      "x-ms-discriminator-value": "Microsoft.ServiceBus/Queue",
+      "allOf": [
+        {
+          "$ref": "#/definitions/OutputDataSource"
+        }
+      ],
+      "properties": {
+        "properties": {
+          "$ref": "#/definitions/ServiceBusQueueOutputDataSourceProperties",
+          "description": "The properties that are associated with a Service Bus Queue output. Required on PUT (CreateOrReplace) requests.",
+          "x-ms-client-flatten": true
+        }
+      }
+    },
+    "ServiceBusQueueOutputDataSourceProperties": {
+      "description": "The properties that are associated with a Service Bus Queue output.",
+      "allOf": [
+        {
+          "$ref": "./inputs.json#/definitions/ServiceBusDataSourceProperties"
+        }
+      ],
+      "properties": {
+        "queueName": {
+          "type": "string",
+          "description": "The name of the Service Bus Queue. Required on PUT (CreateOrReplace) requests."
+        },
+        "propertyColumns": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "A string array of the names of output columns to be attached to Service Bus messages as custom properties."
+        },
+        "systemPropertyColumns": {
+          "type": "object",
+          "additionalProperties": {
```

Comment:
>additionalProperties [](http://example.com/codeflow?start=11&length=20)

Highly recommended to specify the property names. These properties are hard to use by consumers of the API. They also will not be discoverable by Azure Policy for support in policy definitions.

Reply:
@mentat9 
Is there a way to enforce this in the swagger? Or do you just mean specify them in the definition? For example, for this property systemPropertyColumns, the object will look something like this:
            "systemPropertyColumns": { 
              "MessageId": "col3", 
              "PartitionKey": "col4"
            }
Where the key can only be certain system properties that we support. I don't think I'm aware of any ways to specify/constrain the possible keys in an object.

Reply:
No, I mean specify them in the object definition. Even specifying some of them is better than none.

ARM recommends against the use of additionalProperties. Open-ended unschematized objects are problematic, even though they appear to make some things easier for RP owners. ARM is evolving toward stronger typing wherever possible because it allows all consumers of your API to reason about it. This has benefits for everyone that interacts with your API from infrastructure, to other Azure features to SDK owners to customers. Many APIs use this, but it's a pattern that should be avoided if at all possible.